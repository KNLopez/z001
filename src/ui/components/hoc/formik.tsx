import { Field as FormikField, FieldProps } from "formik";
import React from "react";
import { Subtract } from "../../form.builder/types/common";
import { FormikHybridFieldProps } from "./utils/types";

export function withFormik<T extends FormikHybridFieldProps, U>(
  Component: React.ComponentType<T>,
) {

  return class extends React.Component<Subtract<T, FieldProps>> {

    public render() {
      const childrenRenderer = (fieldProps: FieldProps<U>) =>
      <Component {...fieldProps} {...this.props as T} />;

      return <FormikField name={this.props.name} render={childrenRenderer}/>;
    }
  };
}
