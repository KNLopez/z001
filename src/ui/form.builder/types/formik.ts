import { FormikConfig, FormikProps } from "formik";
import { Dictionary } from "lodash";
import { FBTemplateSchema } from "./store";

export interface FBFormikInjectedProps {
  render?: (props: FormikProps<any>) => React.ReactNode;
}

export interface FBFormikProps
  extends Omit<FormikConfig<any>, "onSubmit" | "initialValues"> {
  onSubmit?: (values: any) => void;
  initialValues?: any;
  formComponent?: Element;
}

export interface FBDocumentFlowProps {
  id?: string;
  formTemplate?: FBFormTemplateProps;
  formDocument?: FBDocumentFlowProps;
  formInput?: string | Dictionary<any>;
}

export interface FBDocumentProps extends FBDocumentFlowProps {
  // document?: Partial<DocumentRevisionDocument>;
}

export interface FBFormTemplateProps {
  // outputDocumentType: Id;
  schema: FBTemplateSchema[] | string;
}
