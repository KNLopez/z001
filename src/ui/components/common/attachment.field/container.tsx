import { Box } from "@material-ui/core";
import _ from "lodash";
import React from "react";
import FBStore from "../../../form.builder/stores/FBStore";

import { withFormik } from "../../hoc/formik";
import Presenter from "./presenter";
import { AttachmentContainerProps, } from "./utils/types";

export interface Attachment {
  id: string;
  name: string;
  type: string;
  s3link: string;
  s3linkCleanCopy?: string;
  isLoading?: boolean;
}

export const AttachmentFieldContainer: React.FunctionComponent<AttachmentContainerProps> = ({
  name = "",
  component = "button",
  multiple = true,
  field,
  form,
  ...other
}) => {


  const onChange = (event: React.FormEvent<HTMLInputElement>) => {
    let isUserAlerted = false;
    const filteredFiles = Array.from(event.currentTarget.files || []).filter((file) => {
      const type = file.name.split(".").pop();

      if (type === "doc") {
        // TODO: update react-intl and make use of hooks to translate this message
        !isUserAlerted && alert(".doc files are not accepted. Please use .docx");
        isUserAlerted = true;
        return false;
      }
      return true;
    });

    if (filteredFiles.length === 0) {
      return;
    }
    form!.setFieldValue(name, filteredFiles);
    if (FBStore.mode === "form") {
      FBStore.setValues({
        ...FBStore.values,
        [name as string]: _.map(filteredFiles, (attach) => _.pick(attach, ["id", "name", "s3link"])),
      });
    }
    // handleUpload(filteredFiles as File[]);
  };

  return (
    <Box>
      <Presenter
        {...other}
        name={name}
        component={component}
        multiple={multiple}
        onChange={onChange}
        form={form}
        field={field}
      />
    </Box>
  );
};

export default withFormik(AttachmentFieldContainer);
