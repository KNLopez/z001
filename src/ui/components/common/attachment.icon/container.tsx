import { Grid } from "@material-ui/core";
import React, { useState } from "react";
import FBStore from "../../../form.builder/stores/FBStore";
import { withFormik } from "../../hoc/formik";
import PresenterFactory from "./presenter.factory";
import {
  AttachmentIconContainerProps,
} from "./utils/types";
import { Attachment } from "../attachment.field/container";


export let SetFile: React.Context<React.Dispatch<React.SetStateAction<Attachment | undefined>>>;

export const AttachmentIconContainer: React.FunctionComponent<AttachmentIconContainerProps> = ({
  name = "",
  form,
  field,
  ...other
}) => {


  const attachments = field!.value as Attachment[] || [];
  const [selectedFile, setSelectedFile] = useState<undefined | Attachment>(undefined);
  const [anchorEl, setAnchorEl] = useState<null | HTMLElement>(null);
  SetFile = React.createContext(setSelectedFile);



  const handleRemove = (attachment: Attachment) => () => {
    if (!form && !field) { return; }
    const filtered = (field!.value as Attachment[]).filter((a) => a.id !== attachment.id);
    form!.setFieldValue(name, filtered);

    if (FBStore.mode === "form") {
      FBStore.setValues({
        ...FBStore.values,
        [name as string]: filtered,
      });
    }
  };

  const handleContextMenu = (file: Attachment) => (evt: React.MouseEvent<HTMLElement>) => {
    evt.preventDefault();
    setSelectedFile(file);
    setAnchorEl(evt.currentTarget);
  };


  const attachmentsRenderer = attachments.map((attachment) => (
    <PresenterFactory
      {...other}
      key={`attachment-${attachment.id}`}
      name={name}
      attachment={attachment}
      handleRemove={handleRemove}
      form={form}
      field={field}
      handleContextMenu={handleContextMenu}
    />
  ));

  return (
    <Grid container={true} spacing={2}>

      {attachmentsRenderer}
    </Grid>
  );
};

export default withFormik(AttachmentIconContainer);
