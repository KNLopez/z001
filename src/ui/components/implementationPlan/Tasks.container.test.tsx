
import { shallow, ShallowWrapper } from "enzyme";
import toJson from "enzyme-to-json";
import { FormikProps } from "formik";
import React from "react";
import { Props, TasksContainer } from "./Tasks.container";

describe("Tasks container tests", () => {
  let wrapper: ShallowWrapper;
  const formWithTasks = {
    values: {
      implementationPlan: {
        tasks: [{ title: "Title 1" }, { title: "Title 2" }],
      },
    },
  } as FormikProps<any>;
  const form = {
    values: {
      implementationPlan: {
        tasks: [],
      },
    },
  } as FormikProps<any>;
  const defaultProps = {
    remove: jest.fn(),
    push: jest.fn(),
    replace: jest.fn(),
    form,
  } as unknown as Props;
  const getWrapper = (props: Partial<Props>) => shallow(
    <TasksContainer {...defaultProps} {...props} />,
  );

  beforeAll(() => {
    wrapper = getWrapper({});
  });

  it("should match snapshot - with no tasks", () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it("should match snapshot - with 2 tasks", () => {
    const wrapperWithTasks = getWrapper({ form: formWithTasks });
    expect(toJson(wrapperWithTasks)).toMatchSnapshot();
  });
});
