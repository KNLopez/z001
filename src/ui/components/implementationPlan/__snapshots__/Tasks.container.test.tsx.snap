// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tasks container tests should match snapshot - with 2 tasks 1`] = `
<TasksPresenter
  closeDialog={[Function]}
  editTask={[Function]}
  isDialogOpen={false}
  numberOfTasks={2}
  onSubmit={[Function]}
  openDialog={[Function]}
  removeTask={[Function]}
  selectedTask={null}
  tasks={
    Array [
      Object {
        "title": "Title 1",
      },
      Object {
        "title": "Title 2",
      },
    ]
  }
/>
`;

exports[`Tasks container tests should match snapshot - with no tasks 1`] = `
<TasksPresenter
  closeDialog={[Function]}
  editTask={[Function]}
  isDialogOpen={false}
  numberOfTasks={0}
  onSubmit={[Function]}
  openDialog={[Function]}
  removeTask={[Function]}
  selectedTask={null}
  tasks={Array []}
/>
`;
